File: packages/contracts/src/domain/download-queue.ts
// Domain types for the main-process download queue. The queue is an in-memory
// manager in the desktop main process, write-through persisted to a
// `download_queue` table so it survives an app restart; the renderer mirrors it
// in a zustand store, hydrated on mount and kept in sync by a `queue-state`
// event. On restart the main process reloads pending/in-progress rows (resetting
// in-flight items to `queued`) and resumes downloading unless the queue is paused.

/**
 * Download lifecycle status. Tracks the DOWNLOAD only — `done` means the file
 * was written to disk, NOT that it was imported into the library (import is a
 * separate renderer-side concern surfaced via toasts).
 */
export type DownloadQueueStatus =
  | 'queued' // waiting for a concurrency slot
  | 'active' // yt-dlp running, downloading
  | 'converting' // yt-dlp post-processing (ExtractAudio/Merger)
  | 'done' // file written, path known (NOT yet imported)
  | 'error' // yt-dlp failed
  | 'canceled'; // user cancelled (distinct from error)

export interface DownloadQueueItem {
  /** Canonical identity. Generated at enqueue (randomUUID). */
  id: string;
  /** Source URL passed to yt-dlp. Secondary lookup key (search/playlist). */
  url: string;
  /**
   * YouTube video id when known (recommendations). Secondary lookup key +
   * needed for cacheYoutubeId at import time.
   */
  youtubeId?: string;
  /** Display title for the row. */
  title: string;
  /** Artwork URL for the row (from the search/import result), when known. */
  thumbnail?: string;
  status: DownloadQueueStatus;
  /** 0–100. Meaningful for active/converting; 0 when queued. */
  progress: number;
  /** Resolved file path once status === 'done'. */
  filePath?: string;
  /** Translated/raw error message when status === 'error'. */
  error?: string;
  /** Playlist-import batch grouping (absent for single downloads). */
  batchId?: string;
  batchIndex?: number;
  /**
   * Batch intent, denormalized onto every batch item so the renderer can
   * reconstruct the batch coordinator after a restart (import in order +
   * recreate the playlist) — these live only here on disk, not in any renderer
   * store. Absent for single downloads.
   */
  batchSourceTitle?: string | null;
  batchCreatePlaylist?: boolean;
  /** Timestamps (ms epoch) for ordering + clear-completed. */
  enqueuedAt: number;
  startedAt?: number;
  finishedAt?: number;
}

export interface DownloadQueueSnapshot {
  items: DownloadQueueItem[];
  maxConcurrency: number;
  activeCount: number;
  /** Whether the queue is paused (no queued items will be promoted to active). */
  paused: boolean;
}

export interface EnqueueDownloadInput {
  url: string;
  youtubeId?: string;
  title: string;
  thumbnail?: string;
  batchId?: string;
  batchIndex?: number;
  /** Batch intent (see `DownloadQueueItem`) — required together with `batchId`. */
  batchSourceTitle?: string | null;
  batchCreatePlaylist?: boolean;
}

File: crates/shiranami-core/src/models/download_queue.rs
//! The download queue, ported from
//! `packages/contracts/src/domain/download-queue.ts`.
//!
//! The queue is an in-memory manager write-through persisted to a
//! `download_queue` table so it survives a restart; the renderer mirrors it in a
//! zustand store, hydrated on mount and kept in sync by a `queue-state` event.

use serde::{Deserialize, Serialize};
use specta::Type;
use specta_typescript::Number;

/// Download lifecycle status.
///
/// Tracks the **download** only: [`DownloadQueueStatus::Done`] means the file was
/// written to disk, not that it was imported into the library — import is a
/// separate renderer-side concern surfaced via toasts.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, Type)]
#[serde(rename_all = "lowercase")]
pub enum DownloadQueueStatus {
    /// Waiting for a concurrency slot.
    Queued,
    /// yt-dlp running, downloading.
    Active,
    /// yt-dlp post-processing (`ExtractAudio` / `Merger`).
    Converting,
    /// File written, path known. **Not** yet imported.
    Done,
    /// yt-dlp failed.
    Error,
    /// User cancelled — deliberately distinct from [`DownloadQueueStatus::Error`].
    Canceled,
}

/// One item in the download queue.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct DownloadQueueItem {
    /// Canonical identity, generated at enqueue.
    pub id: String,
    /// Source URL passed to yt-dlp. Secondary lookup key for search/playlist.
    pub url: String,
    /// YouTube video id when known; needed to cache the id at import time.
    #[specta(optional)]
    pub youtube_id: Option<String>,
    /// Display title for the row.
    pub title: String,
    /// Artwork URL for the row, when known.
    #[specta(optional)]
    pub thumbnail: Option<String>,
    /// Current lifecycle status.
    pub status: DownloadQueueStatus,
    /// Percentage 0–100. Meaningful while active or converting; `0` when queued.
    #[specta(type = Number)]
    pub progress: f64,
    /// Resolved file path once the status is [`DownloadQueueStatus::Done`].
    #[specta(optional)]
    pub file_path: Option<String>,
    /// Error message when the status is [`DownloadQueueStatus::Error`].
    #[specta(optional)]
    pub error: Option<String>,
    /// Playlist-import batch grouping; absent for single downloads.
    #[specta(optional)]
    pub batch_id: Option<String>,
    /// Position within the batch.
    #[specta(optional)]
    pub batch_index: Option<u32>,
    /// Source playlist title, denormalized onto every batch item so the renderer
    /// can reconstruct the batch coordinator after a restart. This lives only
    /// here on disk, in no renderer store.
    #[specta(optional)]
    pub batch_source_title: Option<String>,
    /// Whether finishing the batch should create a playlist.
    #[specta(optional)]
    pub batch_create_playlist: Option<bool>,
    /// Enqueue time, epoch milliseconds. Orders the queue.
    #[specta(type = Number)]
    pub enqueued_at: i64,
    /// Start time, epoch milliseconds.
    #[specta(optional, type = Option<Number>)]
    pub started_at: Option<i64>,
    /// Finish time, epoch milliseconds. Drives clear-completed.
    #[specta(optional, type = Option<Number>)]
    pub finished_at: Option<i64>,
}

/// The whole queue as the renderer mirrors it.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct DownloadQueueSnapshot {
    /// Every item, in enqueue order.
    pub items: Vec<DownloadQueueItem>,
    /// How many downloads may run at once.
    pub max_concurrency: u32,
    /// How many are running right now.
    pub active_count: u32,
    /// Whether queued items are being promoted to active.
    pub paused: bool,
}

/// Renderer → main payload to enqueue one download.
#[derive(Debug, Clone, PartialEq, Default, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct EnqueueDownloadInput {
    /// Source URL to pass to yt-dlp.
    pub url: String,
    /// YouTube video id when known.
    #[specta(optional)]
    pub youtube_id: Option<String>,
    /// Display title for the row.
    pub title: String,
    /// Artwork URL for the row.
    #[specta(optional)]
    pub thumbnail: Option<String>,
    /// Playlist-import batch grouping.
    #[specta(optional)]
    pub batch_id: Option<String>,
    /// Position within the batch.
    #[specta(optional)]
    pub batch_index: Option<u32>,
    /// Source playlist title. Required together with `batch_id`.
    #[specta(optional)]
    pub batch_source_title: Option<String>,
    /// Whether finishing the batch should create a playlist.
    #[specta(optional)]
    pub batch_create_playlist: Option<bool>,
}
